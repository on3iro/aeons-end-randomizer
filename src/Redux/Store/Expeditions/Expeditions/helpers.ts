import shortid from 'shortid'

import * as types from 'types'
import { byCost } from 'helpers'

import {
  createArrayWithDefaultValues,
  createIdList,
  getRandomEntity,
  createSupply,
} from 'Redux/helpers'

import { BaseConfig, RollBattleConfig, WinConfig, LossConfig } from './types'

export const calcBattleScore = (tries: number) => {
  switch (tries) {
    case 1: {
      return 6
    }

    case 2: {
      return 4
    }

    case 3: {
      return 2
    }

    default: {
      return 0
    }
  }
}

export const determineBaseTier = () => {}

export const generateBattles = (
  variant: types.Variant,
  expeditionId: string
) => {
  const battles = variant.configList.map(
    (config, index): types.Battle => {
      const isFirst = index === 0

      return {
        id: shortid.generate(),
        nemesisTier: config.tier,
        treasure: config.treasure,
        expeditionId,
        status: isFirst ? 'unlocked' : 'locked',
        tries: 0,
      }
    }
  )

  return battles
}

export const createExpeditionConfig = ({
  variant,
  name,
  bigPocketVariant,
  availableMageIds,
  availableCards,
  availableLevel1TreasureIds,
  tiles,
}: BaseConfig) => {
  const mageIds = createIdList(
    availableMageIds,
    createArrayWithDefaultValues(4, 'EMPTY'),
    getRandomEntity
  ).result
  const { gems, relics, spells } = createSupply(availableCards, tiles)
  const gemsByCost = gems.result.sort(byCost)
  const relicsByCost = relics.result.sort(byCost)
  const spellsByCost = spells.result.sort(byCost)

  const supplyIds = [...gemsByCost, ...relicsByCost, ...spellsByCost].map(
    card => card.id
  )

  const startsWithTreasure = variant.configList[0].tier.tier > 1 // TODO should probably use our rollTreasure property
  const treasureIds = startsWithTreasure
    ? createIdList(
        availableLevel1TreasureIds,
        createArrayWithDefaultValues(5, 'EMPTY'),
        getRandomEntity
      ).result
    : []

  const expeditionId = shortid.generate()
  const battles = generateBattles(variant, expeditionId)

  return {
    battles,
    bigPocketVariant,
    expeditionId,
    mageIds,
    name,
    supplyIds,
    treasureIds,
    variantId: variant.id,
  }
}

// Because we always add these cards to an existing array inside our expedition in our store,
// the count of newly added cards decreases by tier.
// If we ever have a variant, that starts higher than tier 2 we have to
// incorporate changes to roll the whole amount of cards per tier!
const getUpgradedBasicNemesisIdsByBattleTier = ({
  availableTier1Ids,
  availableTier2Ids,
  availableTier3Ids,
  battleTier,
}: {
  availableTier1Ids: string[]
  availableTier2Ids: string[]
  availableTier3Ids: string[]
  battleTier: 1 | 2 | 3 | 4
}) => {
  switch (battleTier) {
    case 1: {
      return [] // No upgraded cards are added on tier 1
    }

    case 2: {
      const tier1Ids = createIdList(
        availableTier1Ids,
        createArrayWithDefaultValues(1, 'EMPTY'),
        getRandomEntity
      ).result
      const tier2Ids = createIdList(
        availableTier2Ids,
        createArrayWithDefaultValues(3, 'EMPTY'),
        getRandomEntity
      ).result
      const tier3Ids = createIdList(
        availableTier3Ids,
        createArrayWithDefaultValues(3, 'EMPTY'),
        getRandomEntity
      ).result

      return [...tier1Ids, ...tier2Ids, ...tier3Ids]
    }

    case 3:
    case 4: {
      const tier1Ids = createIdList(
        availableTier1Ids,
        createArrayWithDefaultValues(1, 'EMPTY'),
        getRandomEntity
      ).result
      const tier2Ids = createIdList(
        availableTier2Ids,
        createArrayWithDefaultValues(1, 'EMPTY'),
        getRandomEntity
      ).result
      const tier3Ids = createIdList(
        availableTier3Ids,
        createArrayWithDefaultValues(2, 'EMPTY'),
        getRandomEntity
      ).result

      return [...tier1Ids, ...tier2Ids, ...tier3Ids]
    }

    default: {
      return [] // Should never occur!
    }
  }
}

const rollNewUpgradedNemesisCards = (
  availableUpgradedBasicNemesisCards: types.UpgradedBasicNemesisCard[],
  previousUpgradedBasicNemesisCards: string[],
  nemesisTier: 1 | 2 | 3 | 4
) => {
  const upgradedCardsWithoutPreviousCards = availableUpgradedBasicNemesisCards.filter(
    upgradedCard => !previousUpgradedBasicNemesisCards.includes(upgradedCard.id)
  )

  const tier1AvailableUpgradedNemesisIds = upgradedCardsWithoutPreviousCards
    .filter(card => card.tier === 1)
    .map(card => card.id)
  const tier2AvailableUpgradedNemesisIds = upgradedCardsWithoutPreviousCards
    .filter(card => card.tier === 2)
    .map(card => card.id)
  const tier3AvailableUpgradedNemesisIds = upgradedCardsWithoutPreviousCards
    .filter(card => card.tier === 3)
    .map(card => card.id)

  const upgradedBasicNemesisCardIds = getUpgradedBasicNemesisIdsByBattleTier({
    battleTier: nemesisTier,
    availableTier1Ids: tier1AvailableUpgradedNemesisIds,
    availableTier2Ids: tier2AvailableUpgradedNemesisIds,
    availableTier3Ids: tier3AvailableUpgradedNemesisIds,
  })

  return [...previousUpgradedBasicNemesisCards, ...upgradedBasicNemesisCardIds]
}

export const createBattle = (config: RollBattleConfig) => {
  const nemesisIds = config.availableNemeses
    .filter(
      nemesis => nemesis.expeditionRating === config.battle.nemesisTier.tier
    )
    .map(nemesis => nemesis.id)
    .filter(nemesisId => !config.previousNemeses.includes(nemesisId))

  const nemesisId = createIdList(
    nemesisIds,
    createArrayWithDefaultValues(1, 'EMPTY'),
    getRandomEntity
  ).result[0]

  const upgradedBasicNemesisCardIds = config.battle.nemesisTier.isNewTier
    ? rollNewUpgradedNemesisCards(
        config.availableUpgradedBasicNemesisCards,
        config.previousUpgradedBasicNemesisCards,
        config.battle.nemesisTier.tier
      )
    : config.previousUpgradedBasicNemesisCards

  return {
    battle: { ...config.battle, nemesisId, status: 'before_battle' },
    upgradedBasicNemesisCardIds,
  }
}

const rollNewEntity = (list: string[]): string =>
  createIdList(list, createArrayWithDefaultValues(1, 'EMPTY'), getRandomEntity)
    .result[0]

export const rollWinRewards = (config: WinConfig) => {
  const newTreasures = createIdList(
    config.treasureIds,
    createArrayWithDefaultValues(5, 'EMPTY'),
    getRandomEntity
  ).result

  const newGem = rollNewEntity(config.gemIds)
  const newRelic = rollNewEntity(config.relicIds)
  const newSpell = rollNewEntity(config.spellIds)

  const updatedBattle = {
    ...config.battle,
    rewards: {
      treasure: newTreasures,
      supplyIds: [newGem, newRelic, newSpell],
      mage: undefined, // we explicitely overwrite the mage reward from earlier losses
    },
  }
  return { battle: updatedBattle }
}

export const rollLossRewards = (config: LossConfig) => {
  switch (config.type) {
    case 'mage': {
      const newMage = rollNewEntity(config.mageIds)

      return {
        ...config.battle,
        rewards: {
          treasure: [],
          mage: newMage,
          supplyIds: [],
        },
      }
    }

    case 'gem': {
      const newGem = rollNewEntity(config.gemIds)

      return {
        ...config.battle,
        rewards: {
          treasure: [],
          mage: undefined,
          supplyIds: [newGem],
        },
      }
    }

    case 'relic': {
      const newRelic = rollNewEntity(config.relicIds)

      return {
        ...config.battle,
        rewards: {
          treasure: [],
          mage: undefined,
          supplyIds: [newRelic],
        },
      }
    }

    case 'spell': {
      const newSpell = rollNewEntity(config.spellIds)

      return {
        ...config.battle,
        rewards: {
          treasure: [],
          mage: undefined,
          supplyIds: [newSpell],
        },
      }
    }

    case 'treasure1': {
      const newTreasure = rollNewEntity(config.treasure1Ids)

      return {
        ...config.battle,
        rewards: {
          treasure: [newTreasure],
          mage: undefined,
          supplyIds: [],
        },
      }
    }

    case 'treasure2': {
      const newTreasure = rollNewEntity(config.treasure2Ids)

      return {
        ...config.battle,
        rewards: {
          treasure: [newTreasure],
          mage: undefined,
          supplyIds: [],
        },
      }
    }

    case 'treasure3': {
      const newTreasure = rollNewEntity(config.treasure3Ids)

      return {
        ...config.battle,
        rewards: {
          treasure: [newTreasure],
          mage: undefined,
          supplyIds: [],
        },
      }
    }

    default: {
      return { ...config.battle }
    }
  }
}
