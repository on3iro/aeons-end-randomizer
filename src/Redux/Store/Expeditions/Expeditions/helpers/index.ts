export * from './convertExpeditionToConfig'
export * from './convertExpeditionFromConfig'

export const calcBattleScore = (tries: number) => {
  switch (tries) {
    case 1: {
      return 6
    }

    case 2: {
      return 4
    }

    case 3: {
      return 2
    }

    default: {
      return 0
    }
  }
}
