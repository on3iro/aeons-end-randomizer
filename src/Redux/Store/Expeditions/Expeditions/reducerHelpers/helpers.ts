import { loop, Cmd } from 'redux-loop'
import { set as setToDb } from 'idb-keyval'
import * as types from 'aer-types/types'

import { State } from '../types'
import { actions } from '../actions'

export const EXPEDITIONS_DB_KEY = 'expeditions-1.11.0'

export const updateBattle = (
  state: State,
  battle: types.Battle,
  additionalBattleProps?: {
    status?: types.BattleStatus
    tries?: number
    rewards?: types.Rewards
  }
) => {
  const oldExpedition = state.expeditions[battle.expeditionId]
  const branches = oldExpedition.sequence.branches

  const updatedBranches = {
    ...branches,
    [battle.id]: {
      ...battle,
      ...additionalBattleProps,
    },
  }

  const newState = {
    ...state,
    expeditions: {
      ...state.expeditions,
      [battle.expeditionId]: {
        ...oldExpedition,
        sequence: {
          ...oldExpedition.sequence,
          branches: updatedBranches,
        },
      },
    },
  }

  return loop(
    newState,
    Cmd.run(setToDb, {
      args: [EXPEDITIONS_DB_KEY, newState],
      successActionCreator: actions.setToDBSuccessful,
      failActionCreator: actions.setToDBFailed,
    })
  )
}
