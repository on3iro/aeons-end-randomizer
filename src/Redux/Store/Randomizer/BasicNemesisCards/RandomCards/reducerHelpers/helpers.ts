import seedrandom from 'seedrandom'
import shortid from 'shortid'

import * as types from 'aer-types/types'

import {
  getRandomEntity,
  createSlotList,
  createBasicNemesisCardList,
} from 'Redux/helpers'

type CardMatrix = {
  [key: string]: {
    [key: number]: number
  }
}

export const cardMatrix: CardMatrix = {
  tier1: {
    1: 1,
    2: 3,
    3: 5,
    4: 8,
  },
  tier2: {
    1: 3,
    2: 5,
    3: 6,
    4: 7,
  },
  tier3: {
    1: 7,
    2: 7,
    3: 7,
    4: 7,
  },
}

export const getBasicNemesisCardsByTier = (
  cards: ReadonlyArray<types.BasicNemesisCard>,
  tier: types.NemesisCardTier
) => cards.filter((card) => card.tier === tier)

export const getRandomMinionAmount = (
  minCount: number = 1,
  count: number = 3,
  seed: types.Seed = {
    seed: shortid.generate(),
  }
) => {
  const rng = seedrandom(seed.seed)
  return (
    Math.floor(rng() * (Math.floor(count) - Math.ceil(minCount))) +
    Math.ceil(minCount)
  )
}

export const drawMinions = (
  cards: ReadonlyArray<types.BasicNemesisCard>,
  amount: number,
  seed?: types.Seed
) => {
  const minions = cards.filter((card) => card.type === 'Minion')
  const minionsList = createBasicNemesisCardList(
    minions,
    createSlotList(amount),
    getRandomEntity,
    seed
  ).result

  return minionsList.map((minion) => {
    return {
      id: minion.id,
    }
  })
}

export const drawPowersAndAttacks = (
  cards: ReadonlyArray<types.BasicNemesisCard>,
  amount: number,
  seed?: types.Seed
) => {
  const powersAndAttacks = cards.filter((card) => card.type !== 'Minion')
  const powersAndAttacksList = createBasicNemesisCardList(
    powersAndAttacks,
    createSlotList(amount),
    getRandomEntity,
    seed
  ).result

  return powersAndAttacksList.map((powerAndAttack) => {
    return {
      id: powerAndAttack.id,
    }
  })
}

export const getRandomBasicNemesisCardsByTier = (
  cards: ReadonlyArray<types.BasicNemesisCard>,
  tier: types.NemesisCardTier,
  playerCount: types.PlayerCount,
  seed?: types.Seed
) => {
  // Get cards by tier
  const cardsByTier = getBasicNemesisCardsByTier(cards, tier)

  // Get amount of minions per player
  const minionAmount =
    playerCount === 1 && tier === 1
      ? getRandomMinionAmount(0, 2, seed)
      : getRandomMinionAmount(1, 3, seed)

  // Draw up to max amount of minions from minions
  const randomMinions = drawMinions(cardsByTier, minionAmount, seed)

  // Draw tierCount - minionCount cards from power/attack pool
  const randomPowersAndAttacks = drawPowersAndAttacks(
    cardsByTier,
    cardMatrix[`tier${tier}`][playerCount] - randomMinions.length,
    seed
  )

  return [...randomMinions, ...randomPowersAndAttacks]
}

export const getRandomBasicNemesisCardList = (
  cards: ReadonlyArray<types.BasicNemesisCard>,
  playerCount: types.PlayerCount,
  seed?: types.Seed
) => {
  return {
    Tier1: getRandomBasicNemesisCardsByTier(cards, 1, playerCount, seed),
    Tier2: getRandomBasicNemesisCardsByTier(cards, 2, playerCount, seed),
    Tier3: getRandomBasicNemesisCardsByTier(cards, 3, playerCount, seed),
  }
}
