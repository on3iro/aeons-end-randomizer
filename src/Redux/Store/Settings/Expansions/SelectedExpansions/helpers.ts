import { Expansions } from './types'

const allExpansionsAreSelected = (expansions: Expansions) =>
  Object.values(expansions).every(exp => exp.selected)

export { allExpansionsAreSelected }
