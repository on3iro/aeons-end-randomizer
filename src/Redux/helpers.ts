import * as types from '../types'

type CardListReduceResult = {
  availableCards: types.ICard[]
  result: types.ICard[]
}

/**
 * Given a card object and a setupBlueprint this returns either
 * true or false, depending on the cost of the card and the expectation set by
 * the blueprints operation and cost.
 */
export const filterByCost = (
  card: { cost: number },
  setupBlueprint: {
    operation: types.Operation
    threshold?: number
    values?: Array<number>
  }
) => {
  switch (setupBlueprint.operation) {
    case '<': {
      if (!setupBlueprint.threshold) return true
      return card.cost < setupBlueprint.threshold
    }
    case '>': {
      if (!setupBlueprint.threshold) return true
      return card.cost > setupBlueprint.threshold
    }
    case '<=': {
      if (!setupBlueprint.threshold) return true
      return card.cost <= setupBlueprint.threshold
    }
    case '>=': {
      if (!setupBlueprint.threshold) return true
      return card.cost >= setupBlueprint.threshold
    }
    case '=': {
      if (!setupBlueprint.threshold) return true
      return card.cost === setupBlueprint.threshold
    }
    case 'OR': {
      if (!setupBlueprint.values) return true
      return setupBlueprint.values.indexOf(card.cost) !== -1
    }
    default: {
      return true
    }
  }
}

export const createCardList = (
  availableCards: types.ICard[],
  blueprints: Array<types.IBluePrint>,
  getEntity: <T>(list: Array<T>) => T
): CardListReduceResult =>
  blueprints.reduce(
    (acc: CardListReduceResult, blueprint: types.IBluePrint) => {
      const filteredByCost = acc.availableCards.filter(card =>
        filterByCost(card, blueprint)
      )

      // Skip blueprint if no compatible card is left
      if (filteredByCost.length <= 0) {
        return acc
      }

      const card = getEntity(filteredByCost)

      // Make sure each entity will only be added to the result list once
      const remainingCards = acc.availableCards.filter(
        entity => entity.id !== card.id
      )

      return {
        availableCards: remainingCards,
        result: [...acc.result, card],
      }
    },
    { availableCards, result: [] }
  )

/**
 * Creates a list of randomly selected cards from a given pool of cards by a specific type.
 * The amount of cards generated is determined by the amount of tileSetups matching
 * the given cardType. E.g. if 2 out of 9 tileSetups have the type="Gem" and the
 * specified type is "Gem" as well, a list of two cards will be created.
 *
 * return List<Card>
 */
const getRandomCardsByType = (
  availableCards: ReadonlyArray<types.ICard>,
  tileSetups: ReadonlyArray<types.IBluePrint>,
  cardType: types.CardType
): CardListReduceResult => {
  const cardSlots = tileSetups.filter(({ type }) => type === cardType)
  const availableCardsOfType = availableCards.filter(
    ({ type }) => type === cardType
  )
  return createCardList(availableCardsOfType, cardSlots, getRandomEntity)
}

/**
 * Creates a supply object, containing gems, relics and spells with meta information.
 * The pool of available cards will be determined by the sets the user has selected
 * inside the application settings. The supply will be created via the blueprint
 * of the given tileSetups.
 *
 * return { gems, relics, spells }
 */
export const createSupply = (
  availableCards: ReadonlyArray<types.ICard>,
  tileSetups: ReadonlyArray<types.IBluePrint>
) => {
  const gems = getRandomCardsByType(availableCards, tileSetups, 'Gem')
  const relics = getRandomCardsByType(availableCards, tileSetups, 'Relic')
  const spells = getRandomCardsByType(availableCards, tileSetups, 'Spell')

  return { gems, relics, spells }
}

export const createArrayWithDefaultValues = (
  size: number,
  defaultValue: any
) => {
  return Array.from({ length: size }, () => defaultValue)
}

export const createSlotList = (amount: number): Array<types.IEmptyBluePrint> =>
  createArrayWithDefaultValues(amount, { type: 'EMPTY', operation: 'NoOp' })

type Entity = { id: string } | string

export const generateListFrom = <T extends Entity>(
  availableEntities: Array<T> | ReadonlyArray<T>,
  slots: Array<any>,
  getEntity: <E extends Entity>(list: Array<E>) => E
): { availableEntities: Array<T>; result: Array<T> } => {
  const result = slots.reduce(
    (acc, _) => {
      // If no entity is left, simply return the actual empty slot
      const entity = getEntity(acc.availableEntities)

      if (!entity) {
        return acc
      }

      // Make sure each entity will only be added to the result list once
      const remainingEntities = acc.availableEntities.filter(
        (available: Entity) => {
          if (typeof available === 'string' && typeof entity === 'string') {
            return available !== entity
          }

          if (typeof available === 'object' && typeof entity === 'object') {
            return available.id !== entity.id
          }

          return true
        }
      )

      return {
        availableEntities: remainingEntities,
        result: [...acc.result, entity],
      }
    },
    { availableEntities, result: [] }
  )

  return result
}

export const createTurnOrderCardList = (
  availableCards: types.ITurnOrderCard[],
  slots: types.ITurnOrderCard[],
  getEntity: <T>(list: Array<T>) => T
) => generateListFrom(availableCards, slots, getEntity)

export const createMageList = (
  availableMages: ReadonlyArray<types.Mage>,
  slots: Array<types.Slot>,
  getEntity: <T>(list: Array<T>) => T
) => generateListFrom(availableMages, slots, getEntity)

export const createIdList = (
  availableIds: ReadonlyArray<string>,
  slots: string[],
  getEntity: <T>(list: Array<T>) => T
) => generateListFrom(availableIds, slots, getEntity)

export const shuffleDeck = (
  deck: types.ITurnOrderCard[]
): types.ITurnOrderCard[] => {
  return createTurnOrderCardList(deck, deck, getRandomEntity).result
}

/**
 * Gets a random value from a list. (The wording of entities is just used for semantic context)
 */
export const getRandomEntity = <E>(availableEntities: ReadonlyArray<E>) =>
  availableEntities[Math.floor(Math.random() * availableEntities.length)]

export const getOperationString = (
  operation: types.Operation,
  values?: number[],
  threshold?: number
) => {
  if (operation === 'NoOp') {
    return ''
  }

  if (operation === 'ANY') {
    return operation
  }

  if (operation === 'OR' && values) {
    return values.join('/')
  }

  const thresholdValue = threshold ? threshold : ''

  return `${operation} ${thresholdValue}`
}
