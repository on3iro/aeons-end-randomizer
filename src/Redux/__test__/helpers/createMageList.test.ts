import { createMageList, createSlotList } from '../../helpers'
import { getRandomEntity } from 'Redux/helpers'

describe('createMageList()', () => {
  it('should return same array with getRandomEntity entity getter', () => {
    const availableEntities = [
      {
        id: 'mage1',
        expansion: '',
        name: '',
        uniqueStarters: [],
        mageTitle: '',
        ability: '',
        complexityRating: 1,
        numberOfCharges: 1,
      },
      {
        id: 'mage2',
        expansion: '',
        name: '',
        uniqueStarters: [],
        mageTitle: '',
        ability: '',
        complexityRating: 1,
        numberOfCharges: 1,
      },
      {
        id: 'mage3',
        expansion: '',
        name: '',
        uniqueStarters: [],
        mageTitle: '',
        ability: '',
        complexityRating: 1,
        numberOfCharges: 1,
      },
    ]
    const blueprints = createSlotList(3)

    const result = createMageList(
      availableEntities,
      blueprints,
      getRandomEntity,
      { seed: 'test' }
    )

    expect(result).toMatchSnapshot()
  })

  it('should return modified array with getRandomEntity entity getter', () => {
    const availableEntities = [
      {
        id: 'mage1',
        expansion: '',
        name: '',
        uniqueStarters: [],
        mageTitle: '',
        ability: '',
        complexityRating: 1,
        numberOfCharges: 1,
      },
      {
        id: 'mage2',
        expansion: '',
        name: '',
        uniqueStarters: [],
        mageTitle: '',
        ability: '',
        complexityRating: 1,
        numberOfCharges: 1,
      },
      {
        id: 'mage3',
        expansion: '',
        name: '',
        uniqueStarters: [],
        mageTitle: '',
        ability: '',
        complexityRating: 1,
        numberOfCharges: 1,
      },
      {
        id: 'mage4',
        expansion: '',
        name: '',
        uniqueStarters: [],
        mageTitle: '',
        ability: '',
        complexityRating: 1,
        numberOfCharges: 1,
      },
    ]
    const blueprints = createSlotList(2)

    const result = createMageList(
      availableEntities,
      blueprints,
      getRandomEntity,
      { seed: 'test' }
    )

    expect(result).toMatchSnapshot()
  })

  it('should return smaller list if there are more blueprints than entities', () => {
    const availableEntities = [
      {
        id: 'mage1',
        expansion: '',
        name: '',
        uniqueStarters: [],
        mageTitle: '',
        ability: '',
        complexityRating: 1,
        numberOfCharges: 1,
      },
    ]
    const blueprints = createSlotList(2)

    const result = createMageList(
      availableEntities,
      blueprints,
      getRandomEntity,
      { seed: 'test' }
    )

    expect(result).toMatchSnapshot()
  })
})
