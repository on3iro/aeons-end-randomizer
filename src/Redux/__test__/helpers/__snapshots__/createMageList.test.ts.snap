// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMageList() should return modified array with getRandomEntity entity getter 1`] = `
Object {
  "availableEntities": Array [
    Object {
      "ability": "",
      "complexityRating": 1,
      "expansion": "",
      "id": "mage1",
      "mageTitle": "",
      "name": "",
      "numberOfCharges": 1,
      "uniqueStarters": Array [],
    },
    Object {
      "ability": "",
      "complexityRating": 1,
      "expansion": "",
      "id": "mage3",
      "mageTitle": "",
      "name": "",
      "numberOfCharges": 1,
      "uniqueStarters": Array [],
    },
  ],
  "result": Array [
    Object {
      "ability": "",
      "complexityRating": 1,
      "expansion": "",
      "id": "mage4",
      "mageTitle": "",
      "name": "",
      "numberOfCharges": 1,
      "uniqueStarters": Array [],
    },
    Object {
      "ability": "",
      "complexityRating": 1,
      "expansion": "",
      "id": "mage2",
      "mageTitle": "",
      "name": "",
      "numberOfCharges": 1,
      "uniqueStarters": Array [],
    },
  ],
  "seed": Object {
    "seed": "test",
    "state": Object {
      "S": Array [
        92,
        55,
        235,
        213,
        126,
        59,
        37,
        65,
        117,
        41,
        205,
        7,
        51,
        174,
        214,
        141,
        135,
        1,
        254,
        2,
        230,
        5,
        45,
        42,
        216,
        32,
        88,
        38,
        190,
        146,
        110,
        109,
        215,
        0,
        53,
        128,
        79,
        192,
        152,
        154,
        49,
        218,
        164,
        40,
        35,
        224,
        101,
        242,
        186,
        12,
        248,
        200,
        252,
        142,
        160,
        111,
        249,
        17,
        91,
        147,
        130,
        232,
        96,
        140,
        66,
        54,
        69,
        76,
        75,
        222,
        188,
        149,
        34,
        177,
        184,
        191,
        9,
        178,
        238,
        68,
        217,
        187,
        131,
        189,
        185,
        234,
        194,
        143,
        28,
        121,
        167,
        150,
        100,
        255,
        64,
        106,
        86,
        203,
        204,
        123,
        173,
        170,
        114,
        87,
        195,
        115,
        183,
        19,
        31,
        95,
        193,
        62,
        85,
        105,
        60,
        241,
        180,
        156,
        159,
        208,
        210,
        15,
        244,
        89,
        163,
        108,
        237,
        116,
        3,
        13,
        30,
        129,
        136,
        176,
        112,
        70,
        46,
        61,
        239,
        198,
        165,
        246,
        209,
        196,
        94,
        221,
        107,
        228,
        125,
        207,
        226,
        206,
        63,
        50,
        20,
        52,
        122,
        201,
        145,
        157,
        48,
        148,
        219,
        236,
        181,
        104,
        250,
        139,
        171,
        155,
        43,
        82,
        16,
        166,
        212,
        18,
        134,
        102,
        29,
        168,
        10,
        83,
        151,
        78,
        119,
        93,
        67,
        77,
        133,
        227,
        118,
        197,
        132,
        138,
        175,
        223,
        153,
        26,
        211,
        161,
        144,
        58,
        243,
        23,
        245,
        4,
        74,
        27,
        251,
        24,
        253,
        11,
        80,
        14,
        21,
        162,
        137,
        231,
        240,
        57,
        179,
        71,
        33,
        120,
        124,
        8,
        98,
        97,
        72,
        169,
        247,
        22,
        229,
        25,
        127,
        225,
        233,
        99,
        36,
        73,
        172,
        81,
        182,
        6,
        39,
        56,
        202,
        47,
        90,
        199,
        158,
        220,
        103,
        113,
        44,
        84,
      ],
      "i": 14,
      "j": 148,
    },
  },
}
`;

exports[`createMageList() should return same array with getRandomEntity entity getter 1`] = `
Object {
  "availableEntities": Array [],
  "result": Array [
    Object {
      "ability": "",
      "complexityRating": 1,
      "expansion": "",
      "id": "mage3",
      "mageTitle": "",
      "name": "",
      "numberOfCharges": 1,
      "uniqueStarters": Array [],
    },
    Object {
      "ability": "",
      "complexityRating": 1,
      "expansion": "",
      "id": "mage1",
      "mageTitle": "",
      "name": "",
      "numberOfCharges": 1,
      "uniqueStarters": Array [],
    },
    Object {
      "ability": "",
      "complexityRating": 1,
      "expansion": "",
      "id": "mage2",
      "mageTitle": "",
      "name": "",
      "numberOfCharges": 1,
      "uniqueStarters": Array [],
    },
  ],
  "seed": Object {
    "seed": "test",
    "state": Object {
      "S": Array [
        92,
        55,
        235,
        213,
        126,
        59,
        37,
        65,
        117,
        41,
        205,
        7,
        51,
        174,
        214,
        0,
        171,
        155,
        139,
        1,
        196,
        125,
        45,
        42,
        216,
        32,
        88,
        38,
        190,
        146,
        110,
        109,
        215,
        141,
        53,
        128,
        79,
        192,
        152,
        154,
        49,
        218,
        164,
        40,
        35,
        224,
        101,
        242,
        186,
        12,
        248,
        200,
        252,
        142,
        160,
        111,
        249,
        17,
        91,
        147,
        130,
        232,
        96,
        140,
        66,
        54,
        69,
        76,
        75,
        222,
        188,
        149,
        34,
        177,
        184,
        191,
        9,
        178,
        238,
        68,
        217,
        187,
        131,
        189,
        185,
        234,
        194,
        143,
        28,
        121,
        167,
        150,
        100,
        255,
        64,
        106,
        86,
        203,
        204,
        123,
        173,
        170,
        114,
        87,
        195,
        115,
        183,
        19,
        31,
        95,
        193,
        62,
        85,
        105,
        60,
        241,
        180,
        156,
        159,
        208,
        210,
        15,
        244,
        89,
        163,
        108,
        237,
        116,
        3,
        13,
        30,
        129,
        136,
        176,
        112,
        70,
        46,
        61,
        239,
        198,
        165,
        246,
        209,
        230,
        94,
        221,
        107,
        228,
        5,
        207,
        226,
        206,
        63,
        50,
        20,
        52,
        122,
        201,
        145,
        157,
        48,
        148,
        219,
        236,
        181,
        104,
        250,
        254,
        135,
        2,
        43,
        82,
        16,
        166,
        212,
        18,
        134,
        102,
        29,
        168,
        10,
        83,
        151,
        78,
        119,
        93,
        67,
        77,
        133,
        227,
        118,
        197,
        132,
        138,
        175,
        223,
        153,
        26,
        211,
        161,
        144,
        58,
        243,
        23,
        245,
        4,
        74,
        27,
        251,
        24,
        253,
        11,
        80,
        14,
        21,
        162,
        137,
        231,
        240,
        57,
        179,
        71,
        33,
        120,
        124,
        8,
        98,
        97,
        72,
        169,
        247,
        22,
        229,
        25,
        127,
        225,
        233,
        99,
        36,
        73,
        172,
        81,
        182,
        6,
        39,
        56,
        202,
        47,
        90,
        199,
        158,
        220,
        103,
        113,
        44,
        84,
      ],
      "i": 21,
      "j": 148,
    },
  },
}
`;

exports[`createMageList() should return smaller list if there are more blueprints than entities 1`] = `
Object {
  "availableEntities": Array [],
  "result": Array [
    Object {
      "ability": "",
      "complexityRating": 1,
      "expansion": "",
      "id": "mage1",
      "mageTitle": "",
      "name": "",
      "numberOfCharges": 1,
      "uniqueStarters": Array [],
    },
  ],
  "seed": Object {
    "seed": "test",
    "state": Object {
      "S": Array [
        92,
        55,
        235,
        213,
        126,
        59,
        37,
        65,
        117,
        41,
        205,
        7,
        51,
        174,
        214,
        141,
        135,
        1,
        254,
        2,
        230,
        5,
        45,
        42,
        216,
        32,
        88,
        38,
        190,
        146,
        110,
        109,
        215,
        0,
        53,
        128,
        79,
        192,
        152,
        154,
        49,
        218,
        164,
        40,
        35,
        224,
        101,
        242,
        186,
        12,
        248,
        200,
        252,
        142,
        160,
        111,
        249,
        17,
        91,
        147,
        130,
        232,
        96,
        140,
        66,
        54,
        69,
        76,
        75,
        222,
        188,
        149,
        34,
        177,
        184,
        191,
        9,
        178,
        238,
        68,
        217,
        187,
        131,
        189,
        185,
        234,
        194,
        143,
        28,
        121,
        167,
        150,
        100,
        255,
        64,
        106,
        86,
        203,
        204,
        123,
        173,
        170,
        114,
        87,
        195,
        115,
        183,
        19,
        31,
        95,
        193,
        62,
        85,
        105,
        60,
        241,
        180,
        156,
        159,
        208,
        210,
        15,
        244,
        89,
        163,
        108,
        237,
        116,
        3,
        13,
        30,
        129,
        136,
        176,
        112,
        70,
        46,
        61,
        239,
        198,
        165,
        246,
        209,
        196,
        94,
        221,
        107,
        228,
        125,
        207,
        226,
        206,
        63,
        50,
        20,
        52,
        122,
        201,
        145,
        157,
        48,
        148,
        219,
        236,
        181,
        104,
        250,
        139,
        171,
        155,
        43,
        82,
        16,
        166,
        212,
        18,
        134,
        102,
        29,
        168,
        10,
        83,
        151,
        78,
        119,
        93,
        67,
        77,
        133,
        227,
        118,
        197,
        132,
        138,
        175,
        223,
        153,
        26,
        211,
        161,
        144,
        58,
        243,
        23,
        245,
        4,
        74,
        27,
        251,
        24,
        253,
        11,
        80,
        14,
        21,
        162,
        137,
        231,
        240,
        57,
        179,
        71,
        33,
        120,
        124,
        8,
        98,
        97,
        72,
        169,
        247,
        22,
        229,
        25,
        127,
        225,
        233,
        99,
        36,
        73,
        172,
        81,
        182,
        6,
        39,
        56,
        202,
        47,
        90,
        199,
        158,
        220,
        103,
        113,
        44,
        84,
      ],
      "i": 14,
      "j": 148,
    },
  },
}
`;
